File: 189_rotate_array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        k=k%nums.size();
        for (int i=0;i<nums.size()/2;i++) {
            int temp=nums[i];
            nums[i]=nums[nums.size()-i-1];
            nums[nums.size()-i-1]=temp;
        }
        for (int i=0;i<k/2;i++){
            int temp=nums[i];
            nums[i]=nums[k-1-i];
            nums[k-1-i]=temp;}
        for (int i=k;i<((nums.size()-k)/2+k);i++) {
            int temp=nums[i];
            nums[i]=nums[nums.size()-i+k-1];
            nums[nums.size()-i+k-1]=temp;
        }
        
    }
};

File: algorithms/python3/189_rotate_array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        def helper(nums,left,right):
            mid=(left+right+1)//2
            for i in range(left,mid):
                nums[i],nums[right-i+left]=nums[right-i+left],nums[i]

        k%=len(nums)
        helper(nums,0,len(nums)-1)
        helper(nums,0,k-1)
        helper(nums,k,len(nums)-1)

a=Solution()
nums=[1,2,3,4,5,6,7]
a.rotate(nums,3)
